File: android-mma-recomp/app/src/main/java/com/example/mmarecomp/util/WodParser.kt
package com.example.mmarecomp.util

data class ParsedWodMovement(val nom: String, val quantite: Int?)

/** Parsing best-effort du texte du WOD collé depuis WhatsApp. Ne bloque
 *  jamais la saisie libre : si rien n'est reconnu, le texte brut reste la
 *  donnée de référence (wodContent), le parsing n'est qu'un résumé visuel. */
object WodParser {
    private val knownMovements = linkedMapOf(
        "burpee" to "Burpees", "burpees" to "Burpees",
        "squat" to "Squats", "squats" to "Squats",
        "pompe" to "Pompes", "pompes" to "Pompes", "push-up" to "Pompes", "push up" to "Pompes",
        "fente" to "Fentes", "fentes" to "Fentes", "lunge" to "Fentes",
        "corde a sauter" to "Corde à sauter", "double under" to "Corde à sauter (double-unders)",
        "sprint" to "Sprints",
        "shadow boxing" to "Shadow boxing", "shadow" to "Shadow boxing",
        "clinch" to "Clinch",
        "sac de frappe" to "Sac de frappe", "sac" to "Sac de frappe",
        "medecine ball" to "Med-ball", "med-ball" to "Med-ball", "med ball" to "Med-ball",
        "kettlebell" to "Kettlebell", "kb swing" to "Kettlebell swings",
        "abdo" to "Abdos", "abs" to "Abdos", "gainage" to "Gainage", "plank" to "Gainage",
        "traction" to "Tractions", "tractions" to "Tractions", "pull-up" to "Tractions",
        "box jump" to "Box jump", "jump squat" to "Jump squat",
        "roulade" to "Roulades", "sprawl" to "Sprawls", "sprawls" to "Sprawls",
    )

    private val numberedMovementRegex =
        Regex("""(\d+)\s*(x|reps?|répétitions?)?\s*([a-zàâäéèêëïîôöùûüç\- ]{3,25})""")

    fun parse(text: String): List<ParsedWodMovement> {
        if (text.isEmpty()) return emptyList()
        val normalized = text.lowercase()
        val found = mutableListOf<ParsedWodMovement>()
        val seen = mutableSetOf<String>()

        for (match in numberedMovementRegex.findAll(normalized)) {
            val quantite = match.groupValues[1].toIntOrNull()
            val words = match.groupValues[3]
            for ((keyword, label) in knownMovements) {
                if (words.contains(keyword) && seen.add(label)) {
                    found.add(ParsedWodMovement(label, quantite))
                }
            }
        }

        // Fallback : mot-clé présent n'importe où dans le texte, sans quantité détectée.
        for ((keyword, label) in knownMovements) {
            if (normalized.contains(keyword) && seen.add(label)) {
                found.add(ParsedWodMovement(label, null))
            }
        }

        return found
    }
}

File: ios-mma-recomp/MMARecomp/Utilities/WodParser.swift
import Foundation

struct ParsedWodMovement: Identifiable, Hashable {
    var id = UUID()
    var nom: String
    var quantite: Int?
}

/// Parsing best-effort du texte du WOD collé depuis WhatsApp. Ne bloque
/// jamais la saisie libre : si rien n'est reconnu, le texte brut reste la
/// donnée de référence (wod_content), le parsing n'est qu'un résumé visuel.
enum WodParser {
    private static let knownMovements: [String: String] = [
        "burpee": "Burpees", "burpees": "Burpees",
        "squat": "Squats", "squats": "Squats",
        "pompe": "Pompes", "pompes": "Pompes", "push-up": "Pompes", "push up": "Pompes",
        "fente": "Fentes", "fentes": "Fentes", "lunge": "Fentes",
        "corde a sauter": "Corde à sauter", "double under": "Corde à sauter (double-unders)",
        "sprint": "Sprints",
        "shadow boxing": "Shadow boxing", "shadow": "Shadow boxing",
        "clinch": "Clinch",
        "sac de frappe": "Sac de frappe", "sac": "Sac de frappe",
        "medecine ball": "Med-ball", "med-ball": "Med-ball", "med ball": "Med-ball",
        "kettlebell": "Kettlebell", "kb swing": "Kettlebell swings",
        "abdo": "Abdos", "abs": "Abdos", "gainage": "Gainage", "plank": "Gainage",
        "traction": "Tractions", "tractions": "Tractions", "pull-up": "Tractions",
        "box jump": "Box jump", "jump squat": "Jump squat",
        "roulade": "Roulades", "sprawl": "Sprawls", "sprawls": "Sprawls",
    ]

    static func parse(_ text: String) -> [ParsedWodMovement] {
        guard !text.isEmpty else { return [] }
        let normalized = text.lowercased()
        var found: [ParsedWodMovement] = []
        var seen = Set<String>()

        let pattern = #"(\d+)\s*(x|reps?|répétitions?)?\s*([a-zàâäéèêëïîôöùûüç\- ]{3,25})"#
        if let regex = try? NSRegularExpression(pattern: pattern) {
            let nsrange = NSRange(normalized.startIndex..., in: normalized)
            regex.enumerateMatches(in: normalized, range: nsrange) { match, _, _ in
                guard let match,
                      let numberRange = Range(match.range(at: 1), in: normalized),
                      let wordsRange = Range(match.range(at: 3), in: normalized) else { return }
                let quantite = Int(normalized[numberRange])
                let words = normalized[wordsRange]

                for (keyword, label) in knownMovements where words.contains(keyword) && !seen.contains(label) {
                    seen.insert(label)
                    found.append(ParsedWodMovement(nom: label, quantite: quantite))
                }
            }
        }

        // Fallback : mot-clé présent n'importe où dans le texte, sans quantité détectée.
        for (keyword, label) in knownMovements where normalized.contains(keyword) && !seen.contains(label) {
            seen.insert(label)
            found.append(ParsedWodMovement(nom: label, quantite: nil))
        }

        return found
    }
}
